runtime/to-json-processor.h

```diff
@@ -0,0 +1,199 @@
+  }
+
+  // support of array<Unknown> compilation
+  void process_impl(const Unknown &/*elem*/) noexcept {}
+
+  void process_impl(const mixed &value) noexcept {
+    switch (value.get_type()) {
+      case mixed::type::NUL:
+        add_null_value();
+        break;
+      case mixed::type::BOOLEAN:
+        add_value(value.as_bool());
+        break;
+      case mixed::type::INTEGER:
+        add_value(value.as_int());
+        break;
+      case mixed::type::FLOAT:
+        add_value(value.as_double());
+        break;
+      case mixed::type::STRING:
+        add_value(value.as_string());
+        break;
+      case mixed::type::ARRAY:
+        process_impl(value.as_array());
+        break;
+    }
+  }
+
+  void add_value(const string &value) noexcept {
+    writer_.String(value);
+  }
+
+  void add_value(const JsonRawString &value) noexcept {
+    writer_.RawString(value.str);
+  }
+
+  void add_value(std::int64_t value) noexcept {
+    writer_.Int64(value);
+  }
+
+  void add_value(bool value) noexcept {
+    writer_.Bool(value);
+  }
+
+  void add_value(double value) noexcept {
+    if (std::isnan(value) || std::isinf(value)) {
+      // preserve behaviour like in php json_encode()
+      value = 0.0;
+    }
+    writer_.Double(value);
+  }
+
+  void add_null_value() noexcept {
+    writer_.Null();
+  }
+
+  impl_::JsonWriter &writer_;
+  const bool flatten_class_{false};
+  bool array_as_hashmap_{false};
+  std::size_t depth_{0};
+};
+
+template<class Tag, class T, class U = Unknown>
+void to_json_impl(const class_instance<T> &klass, impl_::JsonWriter &writer, const array<U> &metadata, std::size_t depth = 0) noexcept {
+  if (depth > 64) {
+    JsonEncoderError::msg.assign("allowed depth=64 of json object exceeded");
+    return;
+  }
+
+  if (klass.is_null()) {
+    writer.Null();
+    return;
+  }
+  constexpr bool flatten_class = impl_::IsJsonFlattenClass<T>::value;
+  if constexpr (!flatten_class) {
+    writer.StartObject();
+  }
+
+  ToJsonVisitor<Tag> visitor{writer, flatten_class, depth};
+  if constexpr (!std::is_empty_v<T>) {
+    klass.get()->accept(visitor);
+  }
+
+  for (auto elem : metadata) {
+    const auto &key = elem.get_key().to_string();
+    visitor({key.c_str(), key.size()}, elem.get_value());
+  }
+
+  if constexpr (!flatten_class) {
+    writer.EndObject();
+  }
+}
+
+template<class Tag>
+template<class I>
+void ToJsonVisitor<Tag>::process_impl(const class_instance<I> &instance) noexcept {
+  to_json_impl<Tag>(instance, writer_, {}, depth_);
+}
+
+template<class Tag, class T, class U = Unknown>
+string f$to_json_impl(Tag /*tag*/, const class_instance<T> &klass, bool pretty_print = false, const array<U> &metadata = {}) noexcept {
+  JsonEncoderError::msg = {};
+
+  impl_::JsonWriter writer{pretty_print};
+  // from_json() expects json string to be a json object,
+  // so don't store null here
+  if (klass.is_null()) {
+    writer.StartObject();
+    writer.EndObject();
+    return writer.GetJson();
```

Comment:
But why? For a null object, the return value should be a string "null". It will be decoded back to null object.

Reply:
Fixed with lots of other stuff